### PR TITLE
Feat/covariance over inputs

### PR DESCRIPTION
**Pull request by jkoeller**
This PR adds the ability to model correlations between predictions. It

1. Adds the "naive" "correlation over estimators" computation to the sklearn random forest implementation and associated test
2. Adds support for a correlation matrix to PredictiveDistribution and child classes, including a new class CorrelatedNormalPredictiveDistribution which is returned from the sklearn random forest when input space correlations are desired.

### CODE REVIEW

File: learners/scikit_learn/random_forest_regression_sklearn.py [hunk 1]
```diff
@@ -181,11 +187,14 @@ def fit(self, data: Data) -> "RandomForestRegressionSklearn":
 
         return self
 
-    def apply(self, data: Data) -> NormalPredictiveDistribution:
+    def apply(self, data: Data, compute_corr: bool=False) -> NormalPredictiveDistribution:
```
Comment (jkoeller) [inline learners/scikit_learn/random_forest_regression_sklearn.py:190]: Actually, I think this is not the right way to do it because it breaks the interface for other learners that do not support this argument. So we either need to add this method argument to each learner's .apply() method, or let compute_corr() be a separate method entirely.

Comment (bfolie) [reply learners/scikit_learn/random_forest_regression_sklearn.py:190]: You definitely don't want to break the `apply` interface. It's simplicity is what makes it extensible.

I don't see why you need the `compute_corr` argument in `apply`, though. You already have the `correlations` argument in `__init__`. If it's equal to `None` then this object does not compute correlations. If it's equal to `naive` then it does.

File: learners/scikit_learn/random_forest_regression_sklearn.py [hunk 2]
```diff
@@ -200,14 +211,41 @@ def apply(self, data: Data) -> NormalPredictiveDistribution:
...
             preds = np.asfarray([tree.predict(xpred) for tree in self._model.estimators_])
-            return NormalPredictiveDistribution(
-                mean=np.mean(preds, axis=0), stddev=np.std(preds, axis=0)
-            )
+            if  self._correlations is None:
+                return NormalPredictiveDistribution(
+                    mean=np.mean(preds, axis=0), stddev=np.std(preds, axis=0)
+                )
+            elif self._correlations == "naive":
+                if (data.num_samples > 25000) and not self._force_corr:
```
Comment (bfolie) [inline learners/scikit_learn/random_forest_regression_sklearn.py:224]: Should the limit also depend on how many labels there are?

File: learners/scikit_learn/random_forest_regression_sklearn.py [hunk 3]
```diff
@@ -200,14 +211,41 @@ def apply(self, data: Data) -> NormalPredictiveDistribution:
...
+                    warn(
+                        "Input correlations requested for >25k predictions."
+                        " Corelation matrix will not be computed."
```
Comment (bfolie) [inline learners/scikit_learn/random_forest_regression_sklearn.py:227]: Mention what's problematic about trying to compute the correlation matrix with >25k samples, to help users understand if they can override.

File: learners/scikit_learn/random_forest_regression_sklearn.py [hunk 4]
```diff
@@ -200,14 +211,41 @@ def apply(self, data: Data) -> NormalPredictiveDistribution:
...
-        if self._uncertainties is None:
+        if self._uncertainties is None and self._correlations is None:
             preds = self._model.predict(xpred)
             return DeltaPredictiveDistribution(mean=preds)
         elif self._uncertainties == "naive":
             preds = np.asfarray([tree.predict(xpred) for tree in self._model.estimators_])
-            return NormalPredictiveDistribution(
-                mean=np.mean(preds, axis=0), stddev=np.std(preds, axis=0)
-            )
+            if  self._correlations is None:
```
Comment (bfolie) [inline learners/scikit_learn/random_forest_regression_sklearn.py:219]:
```suggestion
            if self._correlations is None:
```

File: tests/test_random_forest_regression_sklearn.py [hunk 5]
```diff
@@ -50,6 +51,14 @@ def test_RandomForestRegressionSklearn_2():
     stddev = rf.apply(smlb.TabularData(data=np.array([[-2], [0], [2]]))).stddev
     assert stddev[0] > stddev[1] < stddev[2]
 
+    corr = rf.apply(smlb.TabularData(data=np.array([[-1], [0], [1]]))).corr
+    assert corr.shape == (len(mean), len(mean))
+    assert np.allclose(corr, 
+        [[ 1, -0.165, 0.026],
```
Comment (bfolie) [inline tests/test_random_forest_regression_sklearn.py:57]:
```suggestion
        [[1, -0.165, 0.026], 
```

File: tests/test_random_forest_regression_sklearn.py [hunk 6]
```diff
@@ -50,6 +51,14 @@ def test_RandomForestRegressionSklearn_2():
     stddev = rf.apply(smlb.TabularData(data=np.array([[-2], [0], [2]]))).stddev
     assert stddev[0] > stddev[1] < stddev[2]
 
+    corr = rf.apply(smlb.TabularData(data=np.array([[-1], [0], [1]]))).corr
+    assert corr.shape == (len(mean), len(mean))
+    assert np.allclose(corr, 
+        [[ 1, -0.165, 0.026], 
+        [-0.165, 1 , -0.307],
```
Comment (bfolie) [inline tests/test_random_forest_regression_sklearn.py:58]:
```suggestion
        [-0.165, 1, -0.307],
```

**Comment by bfolie** [comment]
Update with master to get the .travis.yml file, which will run the CI check

File: learners/scikit_learn/random_forest_regression_sklearn.py (hunk 2 above)
Comment (jkoeller) [reply learners/scikit_learn/random_forest_regression_sklearn.py:224]: After performing some memory profiling it appears the number of trees has essentially no impact on the amount of memory required. It's entirely determined by the number of samples, because the correlation matrix is size num_samples^2.

File: learners/scikit_learn/random_forest_regression_sklearn.py (hunk 3 above)
Comment (jkoeller) [reply learners/scikit_learn/random_forest_regression_sklearn.py]: Comment added.

**Comment by jkoeller** [comment]
This PR now includes all the changes from the old "uq-ndme-metrics" branch. To keep things conceptually clear this branch probably should have been merged into master before that merge, but oh well.

**Comment by bfolie** [comment]
Is this ready for re-review? Were there any outstanding issues before the recent merge?

**Comment by jkoeller** [comment]
> Is this ready for re-review? Were there any outstanding issues before the recent merge?

Yes, ready for re-review. I believe the only thing I didn't receive explicit feedback from you on was my comment about why if number of predictions > 25k the correlation matrix won't be computed.

**Review by bfolie — APPROVED** [review]
